interp/src/stk_env.rs

```diff
@@ -0,0 +1,809 @@
+        let mut tr = self;
+        for n in 0..levels {
+            tr = tr.smoosh_once();
+        }
+        tr
+    }
+
+    /// For internal use only
+    /// Set [new] as the topmost scope of [self]
+    fn push_scope(&mut self, new: HashMap<K, V>) {
+        let old_head = mem::replace(&mut self.head, new);
+        self.tail = self.tail.push(old_head);
+    }
+
+    /// ```text
+    /// Consumes two Smooshers, which must share a fork point.
+    /// Merges all topmost scopes above the shared fork point, smooshing the resulting
+    /// node onto the fork point. Returns the resulting Smoosher.
+    /// ```
+    /// ## IMPORTANT INVARIANT
+    /// ```text
+    /// The intersection of the bindings of the two branches
+    /// MUST be the empty set! Otherwise, no guarantee for which binding will be included
+    /// in the merge.
+    /// ```
+    /// # Panics
+    /// ```text
+    /// Panics if [self] and [other] do not share a common fork point, or if either is [empty]
+    /// ```
+    /// # Examples
+    /// ## Pictorial Example
+    /// ```text
+    /// Smoosher 1: (A, B, C, F, G) Smoosher 2: (D, E, F, G)          
+    /// [A]
+    ///  |
+    /// [B]        [D]
+    ///  |          |
+    /// [C]        [E]
+    ///  |          |
+    ///   \        /
+    ///    \      /
+    ///      [F]
+    ///       |
+    ///      [G]
+    /// *merge(Smoosher_1, Smoosher_2)
+    /// Returns:
+    ///      [ABCDE merged and smooshed onto F]
+    ///       |
+    ///      [G]
+    /// ```
+    /// ## Code Example
+    /// good:
+    /// ```
+    /// use interp::stk_env::Smoosher;
+    /// let mut a = Smoosher::new();
+    /// a.set("hello", 15);
+    /// a.set("hi!", 1);
+    /// a.new_scope();
+    /// a.set("bye", 0);
+    /// let mut b = a.fork();
+    /// a.set("hi!", 3);
+    /// b.set("bye", 2);
+    /// let c = Smoosher::merge(a, b);
+    /// assert_eq!(*c.get(&"hi!").unwrap(), 3);
+    /// assert_eq!(*c.get(&"bye").unwrap(), 2);
+    /// assert_eq!(*c.get(&"hello").unwrap(), 15);
+    /// ```
+    /// will panic:
+    /// ```should_panic
+    /// use interp::stk_env::Smoosher;
+    /// let mut a = Smoosher::new();
+    /// a.set("hello", 15);
+    /// let mut b = Smoosher::new();
+    /// b.set("hey", 13);
+    /// let c = Smoosher::merge(a, b); //there's no common fork point
+    /// ```
+    pub fn merge(self, other: Self) -> Self {
+        //find shared fork point; if doesn't exist, panic
+        let mut a = self;
+        let mut b = other;
+        if let Some((depth_a, depth_b)) = Smoosher::shared_fork_point(&a, &b) {
+            //smoosh [self] and [other] to right before that point
+            a = a.smoosh(depth_a - 1);
+            b = b.smoosh(depth_b - 1);
+
+            //now do merge
+
+            //get both heads of smooshed a and b
+            let mut a_head = a.head;
+            //merge a_head and b_head.
+            for (k, v) in b.head {
+                a_head.insert(k, v);
+            }
```

Comment:
Insert returns an option of what was held previously, it may be worth ensuring here that the heads are disjoint by panicking if the return is not none

Reply:
i was wondering how to ensure they're disjoint! thanks!